Src/Web/Web.Shared.Net/Implementation/ExceptionHandlersInjector.cs

```diff
@@ -0,0 +1,763 @@
+                DefineAutoInjectedField(typeBuilder);
+                var telemetryClientField = typeBuilder.DefineField(TelemetryClientFieldName, typeof(TelemetryClient), FieldAttributes.Private | FieldAttributes.InitOnly);
+
+                // emit constructor that assigns telemetry client field
+                EmitConstructor(typeBuilder, typeof(TelemetryClient), telemetryClientField, exceptionLoggerBaseCtor);
+
+                // emit Log method 
+                EmitWebApiLog(typeBuilder, exceptionContextType, exceptionGetter, telemetryClientField, exceptionTelemetryCtor, trackExceptionMethod, baseOnLog);
+
+                // create error WebApiExceptionLogger type
+                var exceptionLoggerType = typeBuilder.CreateType();
+
+                // add WebApiExceptionLogger to list of services
+                var exceptionLogger = Activator.CreateInstance(exceptionLoggerType, telemetryClient);
+                addLogger.Invoke(servicesContainer, new[] { iexceptionLoggerType, exceptionLogger });
+            }
+            catch (ResolutionException e)
+            {
+                // some of the required types/methods/properties/etc were not found.
+                // it may indicate we are dealing with a new version of WebApi library
+                // handle it and log here, we may still succeed with MVC injection
+                WebEventSource.Log.InjectionFailed("WebApi", e.ToInvariantString());
+            }
+        }
+
+        /// <summary>
+        /// Emits OnLog method.
+        /// </summary>
+        /// <param name="typeBuilder">MVCExceptionFilter type builder.</param>
+        /// <param name="exceptionContextType">Type of ExceptionContext.</param>
+        /// <param name="exceptionGetter">MethodInfo to get ExceptionLoggerContext.Exception.</param>
+        /// <param name="telemetryClientField">FieldInfo of WebAPIExceptionFilter.telemetryClient.</param>
+        /// <param name="exceptionTelemetryCtor">ConstructorInfo of ExceptionTelemetry.</param>
+        /// <param name="trackException">MethodInfo of TelemetryClient.TrackException(ExceptionTelemetry).</param>
+        /// <param name="baseOnLog">MethodInfo of base (ExceptionLogger) OnLog method.</param>
+        private static void EmitWebApiLog(TypeBuilder typeBuilder, Type exceptionContextType, MethodInfo exceptionGetter, FieldInfo telemetryClientField, ConstructorInfo exceptionTelemetryCtor, MethodInfo trackException, MethodInfo baseOnLog)
+        {
+            // This method emits following code:
+            // public override void OnLog(ExceptionLoggerContext context)
+            // {
+            //    if (context != null && context.Exception != null)
+            //        telemetryClient.TrackException(new ExceptionTelemetry(context.Exception));
+            //    base.OnLog(context);
+            // }
+            // public override void OnLog(ExceptionLoggerContext context)
```

Comment:
Precompiled assemble must still be compiled with some specific version on MVC/WebApi. 
To avoid binding redirects and assembly resolution issues, we considered an option here to use `AppDomain.CurrentDomain.AssemblyResolve` handler to manually resolve assembly versions when it's not possible automatically.
we preferred IL injection approach because
1. such event fires for dlls MVC/WebApi depend on. And while public API surface is stable, particular dll dependencies might change: i.e. new dependency is introduced in MVC v6.
2. it is more invasive and may potentially conflict with other components doing the same.
3. while IL injection is harder to maintain, the thing we are doing here is small and simple and does not seem to be changed in future.

If we'd need to write more complicated logic, i'd say we should change it `AppDomain.CurrentDomain.AssemblyResolve` approach.

Reply:
Before this moment my definition of "small and simple" was a bit different =)

Reply:
:) I mean the code we are emitting is small and simple - just a few lines of code.  I would not try to emit AppInisghts SDK